RealmSwift/SwiftUI.swift

```diff
@@ -645,6 +645,285 @@ extension EnvironmentValues {
+/// and notifies states for the given process
+///
+/// Add AsyncOpen to your ``SwiftUI/View`` or ``SwiftUI/App``,  after a user is already logged,
+/// or if a user is going to be logged
+///
+///     @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen
+///
+/// This will inmeadiatly initiates a `Realm.asyncOpen()` operation which will perform all work needed to get the Realm to
+/// a usable state. (see Realm.asyncOpen() documentation)
+///
+/// This property wrapper will publish states of the current `Realm.asyncOpen()` process like progress, errors and an opened realm,
+/// which can be used to update the view
+///
+///     struct AsyncOpenView: View {
+///         @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen
+///
+///         var body: some View {
+///            switch asyncOpen {
+///            case .notOpen:
+///                ProgressView()
+///            case .open(let realm):
+///                ListView()
+///                   .environment(\.realm, realm)
+///            case .error(_):
+///                ErrorView()
+///            case .progress(let progress):
+///                ProgressView(progress)
+///            }
+///         }
+///     }
+///
+/// This opened `realm` can be later injected to the view as an enviroment value which will be used by our property wrappers
+/// to populate the view with data from the opened realm
+///
+///     ListView()
+///        .environment(\.realm, realm)
+///
+@available(iOS 13.0, macOS 11.0, tvOS 13.0, watchOS 6.0, *)
+@propertyWrapper public struct AsyncOpen: DynamicProperty {
+    @Environment(\.realm) var realm
+    @ObservedObject private var storage = ObservableAsyncOpenStorage()
+
+    private func asyncOpenForUser<T: BSON>(_ user: User, partitionValue: T, configuration: Realm.Configuration) {
+        let userConfig = user.configuration(partitionValue: partitionValue, cancelAsyncOpenOnNonFatalErrors: true)
+        let userSyncConfig = userConfig.syncConfiguration
+        var configuration = configuration
+        configuration.syncConfiguration = userSyncConfig
+        storage.asyncOpen(configuration: configuration)
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+                cancel()
+            } receiveValue: { realm in
+                self.storage.realm = realm
+                self.storage.asyncOpenState = .open(realm)
+            }.store(in: &storage.cancellables)
+    }
+
+    /// :nodoc:
+    public var projectedValue: Published<AsyncOpenState>.Publisher {
+        return storage.$asyncOpenState
+    }
+
+    /// :nodoc:
+    public var wrappedValue: AsyncOpenState {
+        storage.asyncOpenState
+    }
+
+    /**
+     This will cancel any notification from the property wrapper states
+     */
+    public func cancel() {
```

Comment:
Without exposing the propertyWrapper itself in some way, this method is not available to the user.

Reply:
What do you mean by not exposing the property wrapper?

Reply:
The property wrapper, when used as a property wrapper (which should be 99.9% of use cases), does not expose this method in any way.

Reply:
E.g.,:
```swift
struct MyView {
    @AsyncOpen var asyncOpen

    // you cannot call `cancel` when the property is declared this way
}
```

Reply:
but the user can so something like this 
```
.onDisappear {
            asyncOpen.cancel()
        }
```
in case they want to cancel the operation

Reply:
They would have to do:
```swift
.onDisappear {
    _asyncOpen.cancel()
}
```
Which is fine, but notably awkward.

Reply:
yep, you are right, users may not use this that much, but who should provide a way to cancel the subscription to the states in a way